plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/BaseIcebergConnectorTest.java

```diff
@@ -3339,4 +3340,43 @@ public void testOptimizeParameterValidation()
                 "ALTER TABLE nation EXECUTE OPTIMIZE (file_size_threshold => '33s')",
                 "\\QUnable to set procedure property 'file_size_threshold' to ['33s']: Unknown unit: s");
     }
+
+    @Test
+    public void testRepartitioningIsForcedDuringOptimize()
+            throws IOException
+    {
+        String tableName = "test_repartitiong_during_optimize_" + randomTableSuffix();
+        assertUpdate("CREATE TABLE " + tableName + " (key integer, value varchar) WITH (partitioning = ARRAY['value'])");
+        try {
+            // optimize an empty table
+            assertQuerySucceeds("ALTER TABLE " + tableName + " EXECUTE OPTIMIZE");
+            assertThat(getActiveFiles(tableName)).isEmpty();
+
+            assertUpdate("INSERT INTO " + tableName + " VALUES (1, 'one')", 1);
+            assertUpdate("INSERT INTO " + tableName + " VALUES (2, 'one')", 1);
+            assertUpdate("INSERT INTO " + tableName + " VALUES (3, 'one')", 1);
+            assertUpdate("INSERT INTO " + tableName + " VALUES (4, 'one')", 1);
+            assertUpdate("INSERT INTO " + tableName + " VALUES (5, 'one')", 1);
+            assertUpdate("INSERT INTO " + tableName + " VALUES (6, 'one')", 1);
+            assertUpdate("INSERT INTO " + tableName + " VALUES (7, 'one')", 1);
+            assertUpdate("INSERT INTO " + tableName + " VALUES (8, 'two')", 1);
+            assertUpdate("INSERT INTO " + tableName + " VALUES (9, 'two')", 1);
+            assertUpdate("INSERT INTO " + tableName + " VALUES (10, 'three')", 1);
+
+            Set<String> initialFiles = getActiveFiles(tableName);
+            assertThat(initialFiles).hasSize(10);
+
+            computeActual("ALTER TABLE " + tableName + " EXECUTE OPTIMIZE");
```

Comment:
Use session with high `preferred_write_partitioning_min_number_of_partitions` to verify repartitioning is indeed forced, as the test class could have it artificially lowered.